temporalio/contrib/opentelemetry.py

```diff
@@ -0,0 +1,544 @@
+        return _pickle_span(span)
+
+    def _end_workflow_span(self, b: bytes) -> None:
+        _unpickle_span(b, self.tracer).end()
+
+
+class _InputWithMutableHeaders(Protocol):
+    headers: Optional[Mapping[str, temporalio.api.common.v1.Payload]]
+
+
+class _TracingClientOutboundInterceptor(temporalio.client.OutboundInterceptor):
+    def __init__(
+        self, next: temporalio.client.OutboundInterceptor, root: TracingInterceptor
+    ) -> None:
+        super().__init__(next)
+        self.root = root
+
+    async def start_workflow(
+        self, input: temporalio.client.StartWorkflowInput
+    ) -> temporalio.client.WorkflowHandle[Any, Any]:
+        prefix = (
+            "StartWorkflow" if not input.start_signal else "SignalWithStartWorkflow"
+        )
+        with self.root._start_as_current_span(
+            f"{prefix}:{input.workflow}",
+            attributes={"temporalWorkflowID": input.id},
+            input=input,
+        ):
+            return await super().start_workflow(input)
+
+    async def query_workflow(self, input: temporalio.client.QueryWorkflowInput) -> Any:
+        with self.root._start_as_current_span(
+            f"QueryWorkflow:{input.query}",
+            attributes={"temporalWorkflowID": input.id},
+            input=input,
+        ):
+            return await super().query_workflow(input)
+
+    async def signal_workflow(
+        self, input: temporalio.client.SignalWorkflowInput
+    ) -> None:
+        with self.root._start_as_current_span(
+            f"SignalWorkflow:{input.signal}",
+            attributes={"temporalWorkflowID": input.id},
+            input=input,
+        ):
+            return await super().signal_workflow(input)
+
+
+class _TracingActivityInboundInterceptor(temporalio.worker.ActivityInboundInterceptor):
+    def __init__(
+        self,
+        next: temporalio.worker.ActivityInboundInterceptor,
+        root: TracingInterceptor,
+    ) -> None:
+        super().__init__(next)
+        self.root = root
+
+    async def execute_activity(
+        self, input: temporalio.worker.ExecuteActivityInput
+    ) -> Any:
+        info = temporalio.activity.info()
+        with self.root.tracer.start_as_current_span(
+            f"RunActivity:{info.activity_type}",
+            context=self.root._context_from_headers(input.headers),
+            attributes={
+                "temporalWorkflowID": info.workflow_id,
+                "temporalRunID": info.workflow_run_id,
+                "temporalActivityID": info.activity_id,
+            },
+        ):
+            return await super().execute_activity(input)
+
+
+class _WorkflowExternFunctions(TypedDict):
+    __temporal_opentelemetry_start_span: Callable[
+        [_CarrierDict, str, opentelemetry.util.types.Attributes], bytes
+    ]
+    __temporal_opentelemetry_end_span: Callable[[bytes], None]
+
+
+class TracingWorkflowInboundInterceptor(temporalio.worker.WorkflowInboundInterceptor):
```

Comment:
How does this work during replay? What happens if a span is created in one worker and ends in another? Won't we end up with a bunch of orphaned spans since the span ID is not created deterministically?

Reply:
I am not sure how span ID matters. Most of these calls do not create spans on replay. Are you speaking specifically about the `even_on_replay=True` for `execute_workflow`? I basically just copied the code verbatim from the Go interceptor code.

My tests aren't showing any orphaned spans during replay (done via query after completion), but maybe it is not properly testing after eviction? Can you describe the exact test to write to confirm behavior here?